### PR TITLE
Symlink create links should tell when stable implicitly moved

### LINKED ISSUE #392: create_links should tell when it is the stable link that is moved
We are getting messages from create_links like this:

> Existing symlink 2023.06-py38 moved from 2023.06.10-py38 to 2023.06.11-py38

which are nice, but it would be even nicer if it also could state that "stable" is actually pointing to this, and thus stable is impliclitly moved.

### PR DESCRIPTION
**Pull request by jonathan-eq**
Resolves #392 
Depends on #391

### CODE REVIEW

File: komodo/symlink/create_links.py [hunk 1]
```diff
@@ -17,6 +17,13 @@ def working_dir(path):
         os.chdir(prev_dir)
 
 
+def traverse_symlink_tree(key, link_dict):
+    values = [k for k, v in link_dict.items() if v == key]
+    for value in values:
+        traverse_symlink_tree(value, link_dict)
+        print(f"{value} was implicitly moved as a side effect")
```
Comment (berland) [inline komodo/symlink/create_links.py:24]: I think we can drop "as a side effect". This is the main effect that we target after all.

Comment (jonathan-eq) [reply komodo/symlink/create_links.py]: Yes. sounds good!

File: komodo/symlink/create_links.py [hunk 2]
```diff
@@ -43,10 +51,10 @@ def create_symlinks(links_dict):
 
     if not os.path.isdir(root_folder):
         raise ValueError(f"{root_folder} is not a directory or does not exist")
-
+    links_dict_links = links_dict["links"]
     with working_dir(root_folder):
         for dst, src in links_dict["links"].items():
-            _create_link(src, dst, links_dict["links"])
+            _create_link(src, dst, links_dict_links)
```
Comment (berland) [inline komodo/symlink/create_links.py:57]: why this change?

File: komodo/symlink/create_links.py [hunk 3]
```diff
@@ -17,6 +17,13 @@ def working_dir(path):
         os.chdir(prev_dir)
 
 
+def traverse_symlink_tree(key, link_dict):
```
Comment (berland) [inline komodo/symlink/create_links.py:20]: This should be renamed to something that describes what it is doing (locating implicitly moved symlinks), not how it is doing it.

File: komodo/symlink/create_links.py [hunk 4]
```diff
@@ -17,6 +17,13 @@ def working_dir(path):
...
+    values = [k for k, v in link_dict.items() if v == key]
```
Comment (berland) [inline komodo/symlink/create_links.py:21]: (pylint is not fond of single-letter variables..)

File: tests/test_link_io_structure.py [hunk 5]
```diff
@@ -95,7 +95,32 @@ def test_overwrite_symlink_stdout(tmpdir, capsys):
         os.symlink("2023.06", "2023")
         create_symlinks(links_dict)
         captured = capsys.readouterr()
-        assert "Existing symlink 2023 moved from 2023.06 to 2023.07\n" == captured.out
+        print(captured.out)
```
Comment (berland) [inline tests/test_link_io_structure.py:98]: leftover print?

File: komodo/symlink/create_links.py [hunk 6]
```diff
@@ -17,6 +17,13 @@ def working_dir(path):
...
+    for value in values:
+        traverse_symlink_tree(value, link_dict)
+        print(f"{value} was implicitly moved")
```
Comment (berland) [inline komodo/symlink/create_links.py:24]: This means we will get one line of output for every level in the link-tree. Will that lead to too much noise in the slack message? Perhaps we only want the topmost change, or it might be possible to give all the information in one line.

File: komodo/symlink/create_links.py (hunk 2 above)
Comment (jonathan-eq) [reply komodo/symlink/create_links.py:57]: I used the value multiple places in previous attempt, so I had it as a variable instead. Removing the change now.

File: komodo/symlink/create_links.py (hunk 3 above)
Comment (jonathan-eq) [reply komodo/symlink/create_links.py:20]: Yes, agreed

File: tests/test_link_io_structure.py (hunk 5 above)
Comment (jonathan-eq) [reply tests/test_link_io_structure.py:98]: Yes, is it possible to configure pylint to catch leftover print statements in tests?

File: komodo/symlink/create_links.py (hunk 6 above)
Comment (jonathan-eq) [reply komodo/symlink/create_links.py:24]: I don't know how long a symlink-chain usually is, but it would be better to know too much than not enough. I will try out different formattings, maybe we can shorten it without leaving out information.

File: tests/test_link_io_structure.py (hunk 5 above)
Comment (berland) [reply tests/test_link_io_structure.py]: Don't think so.

File: komodo/symlink/create_links.py [hunk 7]
```diff
@@ -17,6 +17,15 @@ def working_dir(path):
         os.chdir(prev_dir)
 
 
+def locate_implicitly_moved_symlinks(key, link_dict):
```
Comment (berland) [inline komodo/symlink/create_links.py:20]: type this function definition?

File: komodo/symlink/create_links.py [hunk 8]
```diff
@@ -30,6 +39,7 @@ def _create_link(src, dst, link_dict):
             return
         os.remove(dst)
         print(f"Existing symlink {dst} moved from {existing_link} to {src}")
+        locate_implicitly_moved_symlinks(dst, link_dict)
```
Comment (berland) [inline komodo/symlink/create_links.py:42]: It is hard to read from the function name that this has a side-effect to do a `print`. I think it should return a string that we can inject into the print statement on the line above (then we can format it inside a paranthesis e.g).

File: komodo/symlink/create_links.py [hunk 9]
```diff
@@ -30,6 +44,9 @@ def _create_link(src, dst, link_dict):
             return
         os.remove(dst)
         print(f"Existing symlink {dst} moved from {existing_link} to {src}")
+        implicitly_moved_sources = get_implicitly_moved_symlinks(dst, link_dict)
+        for source in implicitly_moved_sources:
+            print(f"{source} was implicitly moved")
```
Comment (berland) [inline komodo/symlink/create_links.py:49]: I think I only want the last one, would that the first or last element of the list? And since this is to be sent to a Slack message, I would like to have it formatted as a sentence to read. What happens with the lineshifts from `print` statements is a little hard to predict.

File: tests/test_link_io_structure.py [hunk 10]
```diff
@@ -95,7 +95,30 @@ def test_overwrite_symlink_stdout(tmpdir, capsys):
         os.symlink("2023.06", "2023")
         create_symlinks(links_dict)
         captured = capsys.readouterr()
-        assert "Existing symlink 2023 moved from 2023.06 to 2023.07\n" == captured.out
+        assert (
+            "Existing symlink 2023 moved from 2023.06 to 2023.07"
+            in captured.out.split("\n")
```
Comment (berland) [inline tests/test_link_io_structure.py:100]: This assert now leaves the possibility for garbage on subsequent lines. Can you constrain it to not allow that?

File: tests/test_link_io_structure.py [hunk 11]
```diff
@@ -95,7 +95,30 @@ def test_overwrite_symlink_stdout(tmpdir, capsys):
...
+        os.mkdir("g")
+        os.symlink("d", "a")
+        os.symlink("c", "b")
+        os.symlink("d", "c")
+        os.symlink("e", "d")
+        os.symlink("f", "e")
+        create_symlinks(links_dict)
+        captured = capsys.readouterr()
+        for symlink in list("ab"):
+            assert f"{symlink} was implicitly moved" in captured.out.split("\n")
```
Comment (berland) [inline tests/test_link_io_structure.py:120]: This example with single letters is fine, but it is hard to read and verify for correctness. Can you make another example that is using `stable` -> `stable-py38` -> `2023.09.03` f.ex and assert exactly on the full output?

**Review by berland — CHANGES_REQUESTED** [review]

File: tests/test_link_io_structure.py (hunk 11 above)
Comment (jonathan-eq) [reply tests/test_link_io_structure.py:120]: Yes, certainly. It was easier to read when I had the drawing next to me on my desk

File: komodo/symlink/create_links.py (hunk 8 above)
Comment (jonathan-eq) [reply komodo/symlink/create_links.py]: Yes, the method should just return the name of the symlink that was moved, so we can format it later on.

File: komodo/symlink/create_links.py (hunk 9 above)
Comment (jonathan-eq) [reply komodo/symlink/create_links.py]: I updated it. Can you verify if it looks good now / how you would improve it?

File: komodo/symlink/create_links.py [hunk 12]
```diff
@@ -30,7 +44,15 @@ def _create_link(src, dst, link_dict):
         if existing_link == src:
             return
         os.remove(dst)
-        print(f"Existing symlink {dst} moved from {existing_link} to {src}")
+        implicitly_moved_sources = get_implicitly_moved_symlinks(dst, link_dict)
+        message = (
+            f"Existing symlink {dst} moved from {existing_link} to {src}."
+            + " Some symlinks were implicitly moved due to this."
```
Comment (berland) [inline komodo/symlink/create_links.py:50]: Looking better! I would move the punctuation dot to after the paranthesis. Or now that it is more explicit how it will look in Slack, maybe a colon and then just a comma-separated list without paranthesis.

**Comment by jonathan-eq** [comment]
I found an edge case for when there was only one symlink (no chain), then the symlink move would be mentioned twice.
Added some code to check if this is the case, then pop it from the returned list if it is there.

**Review by berland — APPROVED** [review]
👌🏻
